src/lib/hash.c

```diff
@@ -79,18 +79,37 @@
 #include <rnp/rnp_sdk.h>
 #include <stdio.h>
 
-static pgp_map_t hash_alg_map[] = {
-  {PGP_HASH_MD5, "MD5"},
-  {PGP_HASH_SHA1, "SHA1"},
-  {PGP_HASH_RIPEMD, "RIPEMD160"},
-  {PGP_HASH_SHA256, "SHA256"},
-  {PGP_HASH_SHA384, "SHA384"},
-  {PGP_HASH_SHA512, "SHA512"},
-  {PGP_HASH_SHA224, "SHA224"},
-  {PGP_HASH_SM3, "SM3"},
-  {PGP_HASH_CRC24, "CRC24"},
-  {0x00, NULL}, /* this is the end-of-array marker */
-};
+static const struct hash_alg_map_t {
+    uint8_t     type;
+    const char *name;
+    const char *botan_name;
+} hash_alg_map[] = {{PGP_HASH_MD5, "MD5", "MD5"},
+                    {PGP_HASH_SHA1, "SHA1", "SHA-1"},
+                    {PGP_HASH_RIPEMD, "RIPEMD160", "RIPEMD-160"},
+                    {PGP_HASH_SHA256, "SHA256", "SHA-256"},
+                    {PGP_HASH_SHA384, "SHA384", "SHA-384"},
+                    {PGP_HASH_SHA512, "SHA512", "SHA-512"},
+                    {PGP_HASH_SHA224, "SHA224", "SHA-224"},
+                    {PGP_HASH_SM3, "SM3", "SM3"},
+                    {PGP_HASH_CRC24, "CRC24", "CRC24"}};
+
+/*
+ * @params
+ * array:       array of the structures to lookup
+ * id_field     name of the field to compare against
+ * ret_field    filed to return
+ * id           id to look for
+ * ret          return value
+ */
+#define LOOKUP_ARRAY_BY_ID(array, id_field, ret_field, id, ret) \
```

Comment:
It seems like this doesn't really need to be a macro, and I think if we can easily avoid a macro, we ought to do so (unless it provides some good benefits).

What do you think about something like this?

```diff
diff --git a/src/lib/hash.c b/src/lib/hash.c
index 429aad4..f2fe324 100644
--- a/src/lib/hash.c
+++ b/src/lib/hash.c
@@ -80,7 +80,7 @@
 #include <stdio.h>
 
 static const struct hash_alg_map_t {
-    uint8_t     type;
+    pgp_hash_alg_t type;
     const char *name;
     const char *botan_name;
 } hash_alg_map[] = {{PGP_HASH_MD5, "MD5", "MD5"},
@@ -93,23 +93,14 @@ static const struct hash_alg_map_t {
                     {PGP_HASH_SM3, "SM3", "SM3"},
                     {PGP_HASH_CRC24, "CRC24", "CRC24"}};
 
-/*
- * @params
- * array:       array of the structures to lookup
- * id_field     name of the field to compare against
- * ret_field    filed to return
- * id           id to look for
- * ret          return value
- */
-#define LOOKUP_ARRAY_BY_ID(array, id_field, ret_field, id, ret) \
-    do {                                                        \
-        for (size_t __i = 0; __i < ARRAY_SIZE(array); __i++) {  \
-            if ((array)[__i].id_field == (id)) {                \
-                (ret) = (array)[__i].ret_field;                 \
-                break;                                          \
-            }                                                   \
-        }                                                       \
-    } while (0)
+static const struct hash_alg_map_t *lookup_hash_alg_by_id(pgp_hash_alg_t id) {
+    for (size_t i = 0; i < ARRAY_SIZE(hash_alg_map); i++) {
+        if (hash_alg_map[i].type == id) {
+            return &hash_alg_map[i];
+        }
+    }
+    return NULL;
+}
 
 /**
  * \ingroup Core_Print
@@ -121,17 +112,15 @@ static const struct hash_alg_map_t {
 const char *
 pgp_show_hash_alg(uint8_t hash)
 {
-    const char *ret = NULL;
-    LOOKUP_ARRAY_BY_ID(hash_alg_map, type, name, hash, ret);
-    return ret;
+    const struct hash_alg_map_t* entry = lookup_hash_alg_by_id(hash);
+    return entry ? entry->name : NULL;
 }
 
 const char *
 pgp_hash_name_botan(pgp_hash_alg_t hash)
 {
-    const char *ret = NULL;
-    LOOKUP_ARRAY_BY_ID(hash_alg_map, type, botan_name, hash, ret);
-    return ret;
+    const struct hash_alg_map_t* entry = lookup_hash_alg_by_id(hash);
+    return entry ? entry->botan_name : NULL;
 }
 
 const char *
```

Reply:
I agree on this - we should minimize macros usage for code readability. Except performance cases - where you need to duplicate bunch of code lines to trade memory for speed, and places where there is no other solution - like `RNP_LOG`.
In other cases it is more clear and simple to use a function instead of macros.

Reply:
This macro will be used for not only the one related to ``hash_alg_map_t``. I don't like returning ``void*``

Reply:
Extensions are even worse :) But what is the problem in having another functions named `lookup_encryption_alg_by_id`, `lookup_pk_alg_by_id` which will do the same cycle? It is just another 20 bytes of code, and 4-5 code lines, in price of better readability and usability.
Nobody here loves code duplication but there should be some compromise between complicating things and saving code lines.

Reply:
Yeah I would say code readability is generally more important than reducing duplication (/DRY), particularly when working in a language like C.

Reply:
@ni4 X number of useless functions like that can be eliminated by simple, readable 5 line macro.

Reply:
In price of this you'll need to keep the same format of maps. For instance, the next thing you'll need to add hash length to the hash map, key bit size to the symmetric key algos, and some flags to the pk algo. You'll need to rewrite all these or start thinking about the next universal solution. 
Anyway, it's question of the coding style and everybody has his own opinion based on experience.